src/Microsoft.VisualStudio.Web.CodeGenerators.Mvc/Areas/AreaGenerator.cs

```diff
@@ -0,0 +1,115 @@
+using Microsoft.VisualStudio.Web.CodeGeneration.DotNet;
+using Microsoft.VisualStudio.Web.CodeGenerators.Mvc.Dependency;
+
+namespace Microsoft.VisualStudio.Web.CodeGenerators.Mvc.Areas
+{
+    [Alias("area")]
+    public class AreaGenerator : ICodeGenerator
+    {
+        private static readonly string[] AreaFolders = new string[]
+        {
+            "Controllers",
+            "Models",
+            "Data",
+            "Views"
+        };
+
+        private IServiceProvider _serviceProvider { get; set; }
+        private IApplicationInfo _appInfo { get; set; }
+        private ILogger _logger { get; set; }
+        private IModelTypesLocator _modelTypesLocator { get; set; }
+
+        public AreaGenerator(IApplicationInfo applicationInfo,
+            IServiceProvider serviceProvider,
+            IModelTypesLocator modelTypesLocator,
+            ILogger logger)
+        {
+            if(serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if(applicationInfo == null)
+            {
+                throw new ArgumentNullException(nameof(applicationInfo));
+            }
+
+            if(logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if(modelTypesLocator == null)
+            {
+                throw new ArgumentNullException(nameof(modelTypesLocator));
+            }
+
+            _serviceProvider = serviceProvider;
+            _logger = logger;
+            _appInfo = applicationInfo;
+            _modelTypesLocator = modelTypesLocator;
+        }
+
+        public async Task GenerateCode(AreaGeneratorCommandLine model)
+        {
+            if (model == null)
+            {
+                throw new ArgumentNullException(nameof(model));
+            }
+
+            EnsureFolderLayout(model);
+            //RegisterAreaInStartup(model.Name, startupType);
+
+            var readmeGenerator = ActivatorUtilities.CreateInstance<ReadMeGenerator>(_serviceProvider);
+            try
+            {
+                await readmeGenerator.GenerateReadmeForArea(model.Name);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogMessage($"Failed to generate a readme: {ex.Message}");
+                throw ex.Unwrap(_logger);
+            }
+        }
+
+        /// <summary>
+        /// Creates a folder hierarchy:
+        ///     ProjectDir
+        ///        \ Areas
+        ///            \ AreaName
+        ///                \ Controllers
+        ///                \ Data
+        ///                \ Models
+        ///                \ Views
+        /// </summary>
+        private void EnsureFolderLayout(AreaGeneratorCommandLine model)
+        {
+            var areaBasePath = Path.Combine(_appInfo.ApplicationBasePath, "Areas");
+            if (!Directory.Exists(areaBasePath))
```

Comment:
i saw e2e test, but it covers only one basic scenario when no such folders exist. In unit tests you could verify all conditions here. You just need to use mocked IFileSystem

---
In reply to: [77363965](https://github.com/aspnet/Scaffolding/pull/295#discussion_r77363965) [](ancestors = 77363965)